homeassistant/components/tado/services.py

```diff
@@ -0,0 +1,55 @@
+"""Services for the Tado integration."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers import selector
+
+from .const import (
+    ATTR_CONFIG_ENTRY,
+    ATTR_MESSAGE,
+    ATTR_READING,
+    DATA,
+    DOMAIN,
+    SERVICE_ADD_METER_READING,
+)
+
+_LOGGER = logging.getLogger(__name__)
+SCHEMA_ADD_METER_READING = vol.Schema(
+    {
+        vol.Required(ATTR_CONFIG_ENTRY): selector.ConfigEntrySelector(
+            {
+                "integration": DOMAIN,
+            }
+        ),
+        vol.Required(ATTR_READING, default=0): vol.Coerce(int),
+    }
+)
+
+
+@callback
+def setup_services(hass: HomeAssistant) -> None:
+    """Set up the services for the Tado integration."""
+
+    async def add_meter_reading(call: ServiceCall) -> None:
+        """Send meter reading to Tado."""
+        entry_id: str = call.data[ATTR_CONFIG_ENTRY]
+        reading: int = call.data[ATTR_READING]
+        _LOGGER.info("Add meter reading %s", reading)
+
+        tadoconnector = hass.data[DOMAIN][entry_id][DATA]
+        response: dict | None = await hass.async_add_executor_job(
+            tadoconnector.set_meter_reading, call.data[ATTR_READING]
+        )
+
+        if not response:
+            raise ServiceValidationError("Could not add meter reading")
```

Comment:
I'm thinking of raising the `HomeAssistantError` in the  `set_meter_reading` instead of logging the error, intercepting it there rather then in the service call. That would make the `None` response handling obsolete. What do you think?

Reply:
I would leave it (somebody else might have a different opinion here), but that way at least it's consistent with the rest

Reply:
replaced `ServiceValidationError` with `HomeAssistantError`